File: golang/reduce_array_size_to_the_half.go
package main

import "sort"

/*
Space : O(n)
Time  : O(n log n)
*/

func minSetSize(arr []int) int {
	n, target := len(arr), len(arr)/2
	d := make(map[int]int)

	for _, x := range arr {
		if _, ok := d[x]; ok {
			d[x]++
		} else {
			d[x] = 1
		}
	}

	var temp []int
	for _, v := range d {
		temp = append(temp, v)
	}

	i, count := 0, 0
	sort.Sort(sort.Reverse(sort.IntSlice(temp)))

	for i < n && count < target {
		count += temp[i]
		i++
	}

	return i
}

File: python3/reduce_array_size_to_the_half.py
"""
Space : O(n)
Time  : O(n log n)
"""

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        n = len(arr)
        target = n // 2
        d = {}
        
        for x in arr:
            if x not in d:
                d[x] = 1
            else:
                d[x] += 1
                
        count = 0
        sc = sorted(list(d.values()), reverse=True)
        i = 0
        
        while count < target and i < n:
            count += sc[i]
            i += 1
        
        return i
        